tests/test_parameter.py

```diff
@@ -736,3 +740,264 @@ def output(self):
+    )
+
+    p = luigi_tools.parameter.DataclassParameter(cls_type=A)
+
+    # Note how json serialization converts d keys into strings
+    string = p.serialize(a)
+    assert string == (
+        '{"a": {"a": 1, "b": "2"}, '
+        '"b": {"c": 3.0, "d": 4}, '
+        '"c": {"e": 5, "f": 6}, '
+        '"d": {"7": 8.0, "9": 10.0}}'
+    )
+
+    a_dict = p.parse(string)
+    assert {
+        "a": {"a": 1, "b": "2"},
+        "b": {"c": 3.0, "d": 4},
+        "c": {"e": 5, "f": 6},
+        "d": {"7": 8.0, "9": 10.0},
+    }
+
+    for obj in (a_dict, a):
+        a_obj = p.normalize(obj)
+        for f in dataclasses.fields(a_obj):
+
+            actual = getattr(a_obj, f.name)
+            expected = getattr(a, f.name)
+
+            # all dicts are converted to frozen ordered dicts
+            assert isinstance(actual, luigi.freezing.FrozenOrderedDict)
+
+            # check that dataclass is hashable
+            assert hash(actual)
+            assert dict(actual) == dict(expected)
+
+
+def test_DataclassParameter__nesting():
+    """Test the DataclassParameter with nested dataclass objects."""
+
+    @dataclasses.dataclass(frozen=True, eq=True)
+    class A:
+        a: str
+        b: float
+
+    @dataclasses.dataclass(frozen=True, eq=True)
+    class B:
+        c: A
+        d: float
+
+    @dataclasses.dataclass(frozen=True, eq=True)
+    class C:
+        e: A
+        f: B
+
+    c = C(e=A(a="1", b=2.0), f=B(c=A(a="2", b=3.0), d=4.0))
+
+    p = luigi_tools.parameter.DataclassParameter(cls_type=C)
+
+    string = p.serialize(c)
+    assert string == '{"e": {"a": "1", "b": 2.0}, "f": {"c": {"a": "2", "b": 3.0}, "d": 4.0}}'
+
+    c_dict = p.parse(string)
+    assert c_dict == {"e": {"a": "1", "b": 2.0}, "f": {"c": {"a": "2", "b": 3.0}, "d": 4.0}}
+
+    c_obj = p.normalize(c_dict)
+    assert c_obj == c
+
+    c_obj = p.normalize(c)
+    assert c_obj == c
+
+
+def test_DataclassParameter__Optional__Union():
+    """Test the DataclassParameter with Optional attributes."""
+
+    @dataclasses.dataclass(frozen=True, eq=True)
+    class A:
+        a: int
+        b: typing.Union[int, str]
+        c: typing.Optional[float] = None
+
+    a1 = A(a=1, b=2.0, c=3.0)
+    a2 = A(a=1, b=2.0)
+
+    p = luigi_tools.parameter.DataclassParameter(cls_type=A)
+
+    s1 = p.serialize(a1)
+    assert s1 == '{"a": 1, "b": 2.0, "c": 3.0}'
+    s2 = p.serialize(a2)
+    assert s2 == '{"a": 1, "b": 2.0, "c": null}'
+
+    d1 = p.parse(s1)
+    assert d1 == {"a": 1, "b": 2.0, "c": 3.0}
+    d2 = p.parse(s2)
+    assert d2 == {"a": 1, "b": 2.0, "c": None}
+
+    n1 = p.normalize(a1)
+    n2 = p.normalize(a2)
+    assert n1 == a1
+    assert n2 == a2
+
+    n1 = p.normalize(d1)
+    n2 = p.normalize(d2)
+    assert n1 == a1
+    assert n2 == a2
+
+
+def test_DataclassParameter__Any():
+    """Test the DataclassParameter with typing.Any attributes."""
+
+    @dataclasses.dataclass(frozen=True, eq=True)
+    class A:
+        a: typing.Any
```

Comment:
Does it work if the given argument is not of a scalar built-in type? Like a dict or even another dataclass?

Reply:
Added attributes that Any is replaced with more complex structures.